### PR TITLE
Template Pipeline: defer `when`, nullary defer triggers, `templateRefExtractor`, deps fns, and two-way bindings

### PR DESCRIPTION
**Pull request by dylhunn**
Various fixes in the template pipeline:
- When conditions, such as `@defer (when(expr))`, are now supported
- Nullary forms of defer triggers, such as `@defer (on hover)`, now use the placeholder element as an implicit trigger
- Defer dependency resolver functions are now generated, when a `defer` block has external dependencies
- Emit `templateRefExtractor` as an argument to the `template` instruction when an `ng-template` has local refs
- Two-way bindings with an implicit receiver  are now ingested correctly (e.g. `[(ngModel)]="foo"`, which implicitly assigns `ctx.foo = $event` in the `ngModelChange` listener)
- Source maps on self-closing elements are now generated for the `elementEnd` instruction, which might sometimes be present (e.g. `input` with an event)
- Enable some passing tests

See individual commits for details.

### CODE REVIEW

**Review by mmalerba — APPROVED** [review]
LGTM

File: packages/compiler/src/template/pipeline/src/phases/defer_resolve_targets.ts [hunk 1]
```diff
@@ -58,6 +58,24 @@ export function resolveDeferTargetNames(job: ComponentCompilationJob): void {
...
+          // placeholder block.
+          if (placeholderView === null) {
+            throw new Error('defer on trigger with no target name must have a placeholder block');
+          }
+          const placeholder = job.views.get(placeholderView);
+          if (placeholder == undefined) {
+            throw new Error('AssertionError: could not find placeholder view for defer on trigger');
+          }
+          for (const placeholderOp of placeholder.ops()) {
+            if (ir.hasConsumesSlotTrait(placeholderOp)) {
```
[the remark this answers is not in the document]
Comment (dylhunn) [reply packages/compiler/src/template/pipeline/src/phases/defer_resolve_targets.ts:71]: Actually, it's also valid for this to be a template, or a repeater, a content, or a projection op. I will update the check.

**Comment by dylhunn** [comment]
Caretaker: g3 presubmit passed in https://critique.corp.google.com/cl/579058797

**Comment by atscott** [comment]
This PR was merged into the repository by commit 699ae60df22b2e2730c32da83ac8632b41f7568c.

**Pull request closed without merging**